### ISSUE TITLE
select name from main.users; is very slow.

### ISSUE THREAD

**Issue by Alkeryn** [scylladb/scylla#10169]
Hey so i made a table as such : 
```cql
CREATE TABLE main.users ( name text primary key) WITH caching = {'enabled': 'true'};
INSERT INTO main.users (name) VALUES ('name')
INSERT INTO main.users (name) VALUES ('name1')
INSERT INTO main.users (name) VALUES ('name2)
INSERT INTO main.users (name) VALUES ('name3)
```
where main is the keyspace.

upon doing the query `select name from users.main;`, on a rust backend it can do only about a hundred request per seconds.
if instead i use the query `select name from users.main where name in ('name','name1','name2','name3');`
i can then do over 100k queries per seconds.

how come trying to get a whole column is that slow ?

i tried using tracing in cql and the former query was almost 200x slower than the later.
even cassandra was much faster and didn't had such an issue, it seems that for the slow query scylla does a lot of "scanning cache for range" 

how could i just get the list of all users in an efficient manner and is it a bug or me missusing the database ?

**Comment by avelanarius** [comment]
@Alkeryn Please post your Rust code that does this testing and your Scylla version, configuration.

I've not been able to reproduce your findings. Tested it with Scylla 4.5.2 in Docker and Java Driver 3.11.0.0. With `select name from main.users` I get the following performance:
```
Time took to perform 200000 queries: 7.515 s - 26613 queries per second
```

With `select name from main.users where name in ('name','name1','name2','name3')` I get the following performance:
```
Time took to perform 200000 queries: 2.841 s - 70397 queries per second
```

Yes, there is some difference, but not the reported several orders-of-magnitude. `select name from main.users` has OK performance. Scylla configured outside Docker and properly tuned could achieve better performance.

Note that, to properly test the query performance of Scylla you should use prepared queries and perform many queries concurrently (not slowly one-by-one) - that's what I did. Attaching the Java program I wrote:  [SelectTests.java.txt](https://github.com/scylladb/scylla/files/8180305/SelectTests.java.txt)

cc @psarna

**Comment by psarna** [comment]
Exact Scylla version would be especially useful, since "select name from users.main" is a full scan, and its implementation changed lately (https://github.com/scylladb/scylla/pull/9209).

**Comment by psarna** [comment]
Wait, never mind, scratch that, the implementation only changed for COUNT(*). I'll try to reproduce the issue locally tomorrow with Rust.

**Comment by Alkeryn** [comment]
@avelanarius this performance difference was taken directly from the time the tracing feature give on the cql shell `tracing on`

but i did also do a benchmark on rust with actix web and on node with fastify
here is the following code for both :

```rust
pub static mut Session : Option<scylla::Session> = None;
pub static mut Prep : Option<scylla::prepared_statement::PreparedStatement> = None;

pub async fn scylla() -> std::result::Result<impl Responder,Box<dyn std::error::Error>> {

    let session : &scylla::Session;
    let prep : &scylla::prepared_statement::PreparedStatement;
    unsafe {
        session = Session.as_ref().unwrap();
        prep = Prep.as_ref().unwrap();
    }


    let mut arr : Vec<String> = Vec::new();
    let res = session.execute(prep, ()).await?;

    if let Some(rows) = res.rows {
        for row in rows {
            arr.push(row.into_typed::<(String,)>()?.0);
        }
    }
    Ok(HttpResponse::Ok().body(format!("{:?}", arr)))
}
```
javascript :
```javascript
const cassandra = require('cassandra-driver')
 
const cluster = new cassandra.Client({
    contactPoints: ["13.36.27.45", "15.236.110.10", "15.236.150.182"],
    localDataCenter: 'AWS_EU_WEST_3',
    credentials: {username: 'scylla', password: 'VM8al7kA5LFnSmt'},
    // keyspace: 'your_keyspace'
})
 
async function routes (fastify, options) {
 
  fastify.get('/', async (request, reply) => {
    const results = await cluster.execute('SELECT name FROM main.users')
    arr = []
    results.rows.forEach(row => arr.push(row.name))
    return arr
  })
    
}
```

**Comment by Alkeryn** [comment]
![image](https://user-images.githubusercontent.com/11599075/156642281-058b299b-2a72-40bb-8533-2851055d0cdb.png)
here you can see the performance difference using bombardier on the js version, where one use `select name from main.users`
and the other uses `select name from main.users where name = 'name1'`

**Comment by psarna** [comment]
Ok, the code samples help a lot, thanks. Full scans are generally a heavy operation, so it's best to parallelize it into multiple scans over smaller ranges, as this blog post explains: https://www.scylladb.com/2017/02/13/efficient-full-table-scans-with-scylla-1-6/

@avelanarius is absolutely right that concurrency is the key for getting top performance from Scylla, and parallelizing the scan would help a lot.

As for your Rust code, one more remark: `execute` is unpaged, and paging is extremely important for queries that can return lots of data - like full scans. Please use `execute_iter` instead, which uses paging: http://rust-driver.docs.scylladb.com/stable/queries/paged.html

**Comment by Alkeryn** [comment]
@psarna thanks for the reply !
i know regarding paging but that was just a code sample to benchmark it somehow, there are only 4 entries so i'm surprised it is that slow, my guess is that scylla scans partitions that are unused.

the performance difference is something like 30x on a local node and jumps to something like 200x on the free trial cluster.
it is snapingly fast when using a where clause but i'm just wondering of a good way to get all users efficiently appart from giving them all a partition number as primary key and using that with the where clause.

**Comment by Alkeryn** [comment]
As for the scylla version, on the local node it is scylla 5.0 running in a lxc container
for the cluster one it is the free trial on scylla cloud.
idk what version they ship.

**Comment by psarna** [comment]
> @psarna thanks for the reply ! i know regarding paging but that was just a code sample to benchmark it somehow, there are only 4 entries so i'm surprised it is that slow, my guess is that scylla scans partitions that are unused.

Yes, that's how it generally works, each partition must be polled in order to check if it's indeed empty (although most of that would usually be cache/index scans, not sstable reads). Full scans are expensive even if they actually return just a few rows.

> 
> the performance difference is something like 30x on a local node and jumps to something like 200x on the free trial cluster. it is snapingly fast when using a where clause but i'm just wondering of a good way to get all users efficiently appart from giving them all a partition number as primary key and using that with the where clause.

You could do that, but partition keys actually already have a "partition number" - the token. It's a 64-bit integer underneath, so you can split one big scan into multiple smaller range scans, like in the blog post I linked.

But, in general, Scylla works best for point queries and range queries, so very frequent full scans could be a strain for the cluster.

> 
> also note that for the rust code i set the prepared statement global variable in the main function in an unsafe block before running the server.

**Comment by psarna** [comment]
@Alkeryn I'm not actually able to reproduce the slowdown you experienced... can you also provide your keyspace definition (I'm mostly interested in the replication factor), your consistency level used (e.g. Rust driver's default is LocalQuorum, while for cqlsh and other drivers it's probably One), and the sample tracing outputs from your cqlsh, showing that one query is 200x slower than the other? Knowing all the configuration and seeing all the tracing details would be really helpful.

**Comment by Alkeryn** [comment]
@psarna hey sorry for the late reply, i got quite busy with life and forgot all about it.
anyway here is a tar.gz of various log files.
to not that there is no such issue using cassandra !
[issue.tar.gz](https://github.com/scylladb/scylla/files/9084474/issue.tar.gz)
